abstract vectorized reduction utils on CPU

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83727
* __->__ #92284

This PR abstracts some reduction utils on CPU, which can be shared by multiple reduction operators, such as `scatter_reduce`, `segment_reduce`, `spmm_reduce`.

No functional change or performance change.



cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
abstract vectorized reduction utils on CPU

[ghstack-poisoned]

mingfeima committed:
Update on "abstract vectorized reduction utils on CPU"


This PR abstracts some reduction utils on CPU, which can be shared by multiple reduction operators, such as `scatter_reduce`, `segment_reduce`, `spmm_reduce`.

No functional change or performance change.



cc jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "abstract vectorized reduction utils on CPU"


This PR abstracts some reduction utils on CPU, which can be shared by multiple reduction operators, such as `scatter_reduce`, `segment_reduce`, `spmm_reduce`.

No functional change or performance change.



cc jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "abstract vectorized reduction utils on CPU"


This PR abstracts some reduction utils on CPU, which can be shared by multiple reduction operators, such as `scatter_reduce`, `segment_reduce`, `spmm_reduce`.

No functional change or performance change.



cc jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "abstract vectorized reduction utils on CPU"


This PR abstracts some reduction utils on CPU, which can be shared by multiple reduction operators, such as `scatter_reduce`, `segment_reduce`, `spmm_reduce`.

No functional change or performance change.



cc jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/cpu/ReduceUtils.h:
```diff
@@ -0,0 +1,158 @@
+#include <c10/util/irange.h>
+
+namespace at::native {
+inline namespace CPU_CAPABILITY {
```
huh, didn't know this was a thing

ezyang reviewed (approved):
okey dokey

jgong5 commented on aten/src/ATen/native/cpu/ReduceUtils.h:
```diff
@@ -0,0 +1,158 @@
+  } else if (reduce == ReductionType::MAX) {
+    return _max(x, y);
+  } else {
+    return _min(x, y);
```
assert `ReductionType::MIN` here?

mingfeima replied:
fixed!

jgong5 commented on aten/src/ATen/native/cpu/ReduceUtils.h:
```diff
@@ -0,0 +1,158 @@
+
+// for Max and Min, propagate NaN:
+template <typename T, ReductionType reduce>
+inline T update(const T& x, const T& y) {
```
The name `update` sounds confusing since it only does a binary op, not update the reduced value. Also, can we make it a local lambda function inside `inline void update(scalar_t* out, scalar_t* data, int64_t K)` if it is not used from other places?

mingfeima replied:
this is also used in other kernels:
* if we do rowwise reduction, then use `void update(scalar_t*,  scalar_t* , int64_t)`
* if we do blocking, then use `void update(Vec&, Vec&)` in the block

jgong5 commented on aten/src/ATen/native/cpu/ReduceUtils.h:
```diff
@@ -0,0 +1,158 @@
+}
+
+template <typename scalar_t, ReductionType reduce>
+inline void write(scalar_t* out, int64_t count, int64_t K) {
```
Since it is an abstraction for reuse, can we add a comment on how the client code can use these functions, e.g., invoking "init - update - write"?

jgong5 replied:
Also, is "init", "update", "write" the name too general, considering it is under "at::native"?

jgong5 commented on aten/src/ATen/native/cpu/ReduceUtils.h:
```diff
@@ -0,0 +1,158 @@
+}
+
+template <typename scalar_t, ReductionType reduce>
+inline void update(scalar_t* out, scalar_t* data, int64_t K) {
```
"reduction combine" is better than "update"? Up to you.

mingfeima replied:
The naming is followed by torch-sparse, i personally have no pereference.

mingfeima committed:
Update on "abstract vectorized reduction utils on CPU"


This PR abstracts some reduction utils on CPU, which can be shared by multiple reduction operators, such as `scatter_reduce`, `segment_reduce`, `spmm_reduce`.

No functional change or performance change.



cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@pytorchbot merge
